sail/impl/plugin.py

```diff
@@ -0,0 +1,394 @@
+                    raise TypeError(
+                        "Please provide the event type either as an argument to the decorator "
+                        "or as a type hint."
+                    )
+
+            else:
+                if typing.get_origin(annotation) in _UNIONS:
+                    resolved_types = {
+                        class_utils.strip_generic(ann) for ann in typing.get_args(annotation)
+                    }
+                else:
+                    resolved_types = {class_utils.strip_generic(annotation)}
+
+                if event_types and resolved_types != set(event_types):
+                    raise TypeError(
+                        "Please make sure the event types provided to the decorator match those "
+                        "provided as a typehint."
+                    )
+
+            for event_type in resolved_types:
+                self.subscribe(event_type, callback)
+
+            return callback
+
+        return wrapper
+
+    async def load(
+        self,
+        client: velum.GatewayClient,
+        command_manager: command_manager_trait.CommandManager,
+    ) -> None:
+
+        self.logger.info("Attempting to load plugin %r...", self.name)
+
+        # Set state...
+        self._client = client
+        self._command_manager = command_manager
+
+        # Load commands...
+        for command in self._commands.values():
+            command_manager.add_command(command)
+
+        # Load listeners...
+        for event, listeners in self._listeners.items():
+            for listener in listeners:
+                client.subscribe(event, listener)
+
+        self.logger.info("Successfully loaded plugin %r!", self.name)
+
+    async def unload(
+        self,
+        client: velum.GatewayClient,
+        command_manager: command_manager_trait.CommandManager,
+    ) -> None:
+
+        # Should realistically always be the case.
+        assert self._client is client
+        assert self._command_manager is command_manager
+
+        self.logger.info("Attempting to unload plugin %r", self.name)
+
+        # Unload commands...
+        for command in self._commands.values():
+            self.command_manager.remove_command(command)
+
+        # Unload listeners...
+        for event, listeners in self._listeners.items():
+            for listener in listeners:
+                self.client.unsubscribe(event, listener)
+
+        # Remove state
+        self._client = None
+        self._command_manager = None
+
+        self.logger.info("Successfully unloaded plugin %s", self.name)
+
+    @property
+    def entrypoints(self) -> typing.Tuple[_EntrypointCallback, _EntrypointCallback]:
+        """Returns the plugin's most basic entrypoints. For use in an extension as
+
+        .. code-block:: python3
+
+            plugin = Plugin()
+
+            ...
+
+            load, unload = plugin.entrypoints
+
+        such that the plugin can be loaded as part of an extension.
```

Comment:
It's a little awkward to read this sentence since it's split into two because of the example. Instead, you should show the example after the first sentence.

Reply:
Makes sense. I fixed this in fd9e199990d85012d99a5a1fbd4884633830c03c.